src/framework-ondata.js

```diff
@@ -1,18 +1,294 @@
+
+    // Do we have a new state? 
+    if ( lastLm !== state.lm ) { 
+
+        // Got any edges out of this port? 
+        if (port.listAttached().length) {
+            port.send( state );
+            port.disconnect();
+            return true;
+        }  
+    } 
+
+    return false;
+}
+
+/**
+ * Set the output state error flag to true if there is a current error, false if not.
+ * 
+ * @param vni 
+ * @param errorFlag (optional) if not undefined, the errorFlag value will be set
+ */
+function setOutputErrorFlag( vni, errorFlag ) {
+
+    var outputState  = vni.outputState();
+    if ( _.isUndefined( errorFlag ) ) { 
+        outputState.error = ! _.isUndefined( vni.errorState().data );
+    } else { 
+        outputState.error = errorFlag;
+    }
+
+    if ( outputState.error ) { 
+         // If we have an error, we should forward the output so downstream nodes know about it
+         outputState.lm = createLm();
+    }
+
+    vni.outputState( outputState );
+}
+
+/** 
+ * Checks if the state was changed since the last state snapshot & if so, 
+ * updates the LM to ensure the change propagates.
+ *
+ * @param stateFacade a getter/setterfunction for the current state
+ * @param lastState a saved clone of the last state 
+ *
+ * @return returns true if state was updated, false if not.
+ */
+function stateChange( stateFacade, lastState ) { 
+
+    var currentState = stateFacade();
+
+    // If got new data or an error state
+    var lastData = (lastState &&  _.isObject( lastState )  ) ? lastState.data : lastState;
+    if ( currentState.data  !== lastData || currentState.error ) {    
+
+        currentState.lm = createLm();
+        stateFacade( currentState );
+        return true;
+    }
+
+    return false;
+}
+
+/*******************************************************************************/
+/* Support for default shouldRunUpdater policy                                 */
+/* Default policy checks to be sure there is data on all attached input ports. */
+/*******************************************************************************/                 
+
+/** 
+ * Retrieve all attached input ports for the noflo component associated with this
+ * node instance. 
+ *
+ * @param node instance context
+ *
+ * @return an array with the attached noflo port objects
+ */
+function attachedInputPorts( node ) { 
+
+    return _.filter( node.inPorts, 
+                     function( port) { return  port.listAttached().length > 0; });
+}
+
+/**
+ * returns true if there is data on all attached input ports; false if not
+ * 
+ * @param vni 
+ */
+function haveAllInputs( vni ) {
+
+    var attachedInPorts = attachedInputPorts( vni.node );
+
+    // If we have undefined data on any port, return false; if we 
+    // have data on all ports, return true
+    return ! _.some( attachedInPorts, function( port ) { 
+
+        var states = vni.inputStates( port.name );
+            
+        return ( _.isUndefined( states ) || 
```

Comment:
No need to change this now, but please add a TODO comment referencing https://github.com/rdf-pipeline/noflo-rdf-pipeline/issues/34 about also checking state.error .